File: laocai-bot-spring-boot-starter/src/main/java/indi/kyson/laocai/bot/model/segment/UnknownSegment.java
package indi.kyson.laocai.bot.model.segment;

import com.fasterxml.jackson.annotation.JsonIgnore;
import tools.jackson.databind.JsonNode;

/**
 * 未知类型的消息段。
 * <p>
 * 协议新增或变更 segment 类型时不应让整条消息反序列化失败，先原样保留原始节点。
 */
public final class UnknownSegment implements Segment {

    record Data(String type, JsonNode raw) {
        UnknownSegment toSegment() {
            return new UnknownSegment(this);
        }
    }

    private final Data data;

    private UnknownSegment(Data data) {
        this.data = data;
    }

    static UnknownSegment of(String type, JsonNode raw) {
        return new UnknownSegment(new Data(type, raw));
    }

    @Override
    public String getType() {
        return data.type();
    }

    @Override
    public Data getData() {
        return data;
    }

    @JsonIgnore
    public JsonNode getRaw() {
        return data.raw();
    }
}

File: laocai-bot-spring-boot-starter/src/main/kotlin/indi/kyson/laocai/bot/segment/UnknownSegment.kt
package indi.kyson.laocai.bot.segment

import com.fasterxml.jackson.annotation.JsonIgnore
import tools.jackson.databind.JsonNode

/**
 * 未知类型的消息段。
 *
 * 协议新增或变更 segment 类型时不应让整条消息反序列化失败，先原样保留原始节点。
 */
class UnknownSegment private constructor(override val data: Data) : Segment {

    data class Data(val type: String, val raw: JsonNode) {
        fun toSegment(): UnknownSegment = UnknownSegment(this)
    }

    override val type: String
        get() = data.type

    @get:JsonIgnore
    val raw: JsonNode
        get() = data.raw

    companion object {
        internal fun of(type: String, raw: JsonNode): UnknownSegment = UnknownSegment(Data(type, raw))
    }
}
